### DISCUSSION TITLE
Is there a limitation on how to run java applications on Theia machine?

### DISCUSSION THREAD

**Discussion by tsiyona** [eclipse-theia/theia#7864] [General] [answered]
I am running a stand-alone Node.js process which acts as a debug-adapter.
This process listens on a TCP port for debug session requests from Theia.
Before starting to listen on the TCP port, my process spawns a java spring-boot application:

//  Basically: java -jar myApp.jar fully.qualified.package.myMainClass

```
let jarFile: string = '/full/path/to/myApp.jar';
let command = '/full/path/to/bin/java';
let mainClass: string = 'fully.qualified.package.myMainClass';
let args = ['-jar', jarFile, mainClass];
	
const childProcess = cp.spawn(command, args, {cwd: process.cwd()});
childProcess.on('close', (code, signal) => {
	if (signal) {
		 throw new Error(`${command} exited with an unexpected signal: ${signal}`);
	}
	if (code !== 0) {
		throw new Error(`${command} exited with an unexpected code: ${code}`));
	}
});

```
After a while (seconds/minutes/days) my java process receives a SIGKILL signal and dies.

Both debug-adapter and the java process run on the same machine as that of Theia.

Is there a limitation on how to run java processes on Theia machine?

**Comment by vince-fugnitto** [comment]
@tsiyona just for clarification, are your debug-adapter and java process interacting in any way whatsoever with theia that is on your machine?

**Comment by tsiyona** [comment, marked as answer]
The debug adapter interacts with theia 'debug' extension (backend) via a standard tcp socket. All DAP requests/responses are transferred between theia debug extension and my debug adapter on this socket.

The java process does not interact with theia in any way. The java process interacts with the debug adapter on a web-socket.

---

My debug adapter runs inside a docker container in kubernetes. I suspect that this is a memory issue. Linux might be the one terminating my java process in order to free some memory to allow Linux to operate properly. I assigned more memory to my pod and I am monitoring now.

1. I just want to make sure there are no limitations as for how to spawn java processes that run on  theia machine. I noticed some other java processes that run on theia machine and were launched using equinox launcher. I wanted to make sure my "java -jar" approach to launching the process is acceptable.

3. Is there a recommended ram quota I should assign to my java process? The other java processes assigned  between 1G and 1.5G.

**Reply by vince-fugnitto** [reply]
> 1. I just want to make sure there are no limitations as for how to spawn java processes that run on  theia machine. I noticed some other java processes that run on theia machine and were launched using equinox launcher. I wanted to make sure my "java -jar" approach to launching the process is acceptable.

I'm not aware of any limitation on running Java processes simultaneously with Theia. Based on your description, this seems to be an issue related to the environment setup more than anything else, perhaps a memory limitation regarding your docker container.

> 2. Is there a recommended ram quota I should assign to my java process? The other java processes assigned  between 1G and 1.5G.

I don't believe we have such recommended quota (at least from Theia), the issue overall does not seem to be Theia-specific. Perhaps others will have some ideas, I'm sorry I'm not able to help further.

**Reply by tsiyona** [reply]
Thank you. I do suspect it is indeed an environment issue.
